File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteMonthlyWeekScheduleRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.firebase.json.MonthlyWeekScheduleJson;
import com.krystianwsul.checkme.firebase.json.ScheduleWrapper;

import junit.framework.Assert;

public class RemoteMonthlyWeekScheduleRecord extends RemoteScheduleRecord {
    RemoteMonthlyWeekScheduleRecord(@NonNull String id, @NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(id, remoteTaskRecord, scheduleWrapper);
    }

    RemoteMonthlyWeekScheduleRecord(@NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(remoteTaskRecord, scheduleWrapper);
    }

    @NonNull
    private MonthlyWeekScheduleJson getMonthlyWeekScheduleJson() {
        MonthlyWeekScheduleJson monthlyWeekScheduleJson = scheduleWrapper.monthlyWeekScheduleJson;
        Assert.assertTrue(monthlyWeekScheduleJson != null);

        return monthlyWeekScheduleJson;
    }

    @Override
    public long getStartTime() {
        return getMonthlyWeekScheduleJson().getStartTime();
    }

    @Nullable
    @Override
    public Long getEndTime() {
        return getMonthlyWeekScheduleJson().getEndTime();
    }

    public int getDayOfMonth() {
        return getMonthlyWeekScheduleJson().getDayOfMonth();
    }

    public int getDayOfWeek() {
        return getMonthlyWeekScheduleJson().getDayOfWeek();
    }

    public boolean getBeginningOfMonth() {
        return getMonthlyWeekScheduleJson().getBeginningOfMonth();
    }

    @Nullable
    public String getCustomTimeId() {
        return getMonthlyWeekScheduleJson().getCustomTimeId();
    }

    @Nullable
    public Integer getHour() {
        return getMonthlyWeekScheduleJson().getHour();
    }

    @Nullable
    public Integer getMinute() {
        return getMonthlyWeekScheduleJson().getMinute();
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(getEndTime() == null);

        getMonthlyWeekScheduleJson().setEndTime(endTime);
        addValue(getKey() + "/monthlyWeekScheduleJson/endTime", endTime);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteMonthlyWeekScheduleRecord.kt
package com.krystianwsul.checkme.firebase.records


import com.krystianwsul.common.firebase.ScheduleWrapper
import com.krystianwsul.common.utils.RemoteCustomTimeId


class RemoteMonthlyWeekScheduleRecord<T : RemoteCustomTimeId> : RemoteScheduleRecord<T> {

    private val monthlyWeekScheduleJson by lazy { createObject.monthlyWeekScheduleJson!! }

    val dayOfMonth by lazy { monthlyWeekScheduleJson.dayOfMonth }

    val dayOfWeek by lazy { monthlyWeekScheduleJson.dayOfWeek }

    val beginningOfMonth by lazy { monthlyWeekScheduleJson.beginningOfMonth }

    override val customTimeId by lazy { monthlyWeekScheduleJson.customTimeId?.let { remoteTaskRecord.getRemoteCustomTimeId(it) } }

    val hour by lazy { monthlyWeekScheduleJson.hour }

    val minute by lazy { monthlyWeekScheduleJson.minute }

    constructor(id: String, remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(id, remoteTaskRecord, scheduleWrapper)

    constructor(remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(remoteTaskRecord, scheduleWrapper)

    override val startTime by lazy { monthlyWeekScheduleJson.startTime }

    override var endTime
        get() = monthlyWeekScheduleJson.endTime
        set(value) {
            if (value == monthlyWeekScheduleJson.endTime)
            return

            monthlyWeekScheduleJson.endTime = value
            addValue("$key/monthlyWeekScheduleJson/endTime", value)
    }

    override fun deleteFromParent() = check(remoteTaskRecord.remoteMonthlyWeekScheduleRecords.remove(id) == this)
}
